sdk/cosmos/azure-cosmos-spark_3/src/main/scala/com/azure/cosmos/spark/CosmosItemsDataSource.scala

```diff
@@ -112,4 +112,182 @@ object CosmosItemsDataSource {
 
     readManyReader.readMany(df.rdd, readManyFilterExtraction)
   }
+
+  /**
+   * Execute transactional batch operations against Cosmos DB.
+   * Operations with the same partition key are executed atomically.
+   * 
+   * Input DataFrame schema:
+   * - Flat columns corresponding to the properties of the Cosmos DB item to be operated on.
+   * - The partition key column name must match the container's partition key path (e.g., "pk" if the path is "/pk").
+   * - The "id" column (String) is required.
+   * - An optional "operationType" column (String) can be provided to specify the operation ("create", "replace", "upsert", "delete", "read") for each row.
+   *   If not provided, the default operation is "upsert".
+   * 
+   * Output DataFrame schema:
+   * - id: String
+   * - partition key column: String (name matches container's partition key path)
+   * - operationType: String
+   * - statusCode: Int
+   * - isSuccessStatusCode: Boolean
+   * - resultDocument: String (optional - for read operations)
+   * - errorMessage: String (optional)
+   * 
+   * @param df DataFrame containing batch operations
+   * @param userConfig Cosmos DB configuration
+   * @return DataFrame with execution results
+   */
+  def writeTransactionalBatch(df: DataFrame, userConfig: java.util.Map[String, String]): DataFrame = {
+    writeTransactionalBatch(df, userConfig, null)
+  }
+
+  /**
+   * Execute transactional batch operations against Cosmos DB with custom output schema.
+   * 
+   * @param df DataFrame containing batch operations
+   * @param userConfig Cosmos DB configuration
+   * @param userProvidedSchema Custom output schema (optional)
+   * @return DataFrame with execution results
+   */
+  def writeTransactionalBatch(df: DataFrame, userConfig: java.util.Map[String, String], userProvidedSchema: StructType): DataFrame = {
+    val batchWriter = new TransactionalBatchWriter(
+      userProvidedSchema,
+      userConfig.asScala.toMap)
+
+    // Initialize and broadcast the client states with partition key definition
+    val (clientMetadataBroadcast, partitionKeyDefBroadcast) = batchWriter.initializeAndBroadcastCosmosClientStatesForContainer()
+    // Deserialize partition key definition from JSON
+    val partitionKeyDefinition = SparkModelBridgeInternal.createPartitionKeyDefinitionFromJson(partitionKeyDefBroadcast.value)
+
+    // Create default extraction function based on DataFrame schema and partition key definition
+    val defaultOperationExtraction = createBatchOperationExtraction(df, partitionKeyDefinition)
+
+    // Execute the batch with the pre-initialized client states
+    batchWriter.writeTransactionalBatchWithClientStates(df.rdd, defaultOperationExtraction, (clientMetadataBroadcast, partitionKeyDefBroadcast))
```

Comment:
I think you have to ensure spark partitioning by PK - and also order by it to have a rpredictabel order (ensure change in pk means you can trigger flushing to backend)?